File: ios/Classes/FlutterTunerPlugin.swift
import Flutter
import UIKit
import AVFoundation

public class FlutterTunerPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  private var eventSink: FlutterEventSink?
  private var tuner: TunerNew?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(name: "flutter_tuner", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(name: "flutter_tuner_stream", binaryMessenger: registrar.messenger())

    let instance = FlutterTunerPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startTuning":
            AVAudioSession.sharedInstance().requestRecordPermission { granted in
                if granted {
                    do {
                        try self.startTuning()
                        result(nil)
                    } catch {
                        result(FlutterError(code: "TUNER_ERROR", message: error.localizedDescription, details: nil))
                    }
                } else {
                    result(FlutterError(code: "401", message: "Microphone permission not granted", details: nil))
                }
            }
        case "stopTuning":
            self.stopTuning()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func startTuning() throws {
        guard tuner == nil else { return }

        tuner = TunerNew { [weak self] frequency in
            DispatchQueue.main.async {
                self?.eventSink?(frequency)
            }
        }

        try tuner?.start()
    }

    private func stopTuning() {
        tuner?.stop()
        tuner = nil
    }
    
  public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    self.eventSink = events
    return nil
  }

  public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    stopTuning()
    self.eventSink = nil
    return nil
  }
}

File: android/src/main/kotlin/com/degastonapps/flutter_tuner/FlutterTunerPlugin.kt
package com.degastonapps.flutter_tuner

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import android.util.Log
import androidx.core.content.ContextCompat
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class FlutterTunerPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
  private lateinit var methodChannel: MethodChannel
  private lateinit var eventChannel: EventChannel
  private lateinit var context: Context
  private val logTag = "Flutter_Tuner"
  private var eventSink: EventChannel.EventSink? = null


  /*
   * This is the function that gets called when the tuner frequency changes.
   */
  private val tunerCallback: (Double) -> Unit = { freq ->
    Handler(Looper.getMainLooper()).post {
      eventSink?.success(freq)
    }
  }

//  private val tuner = Tuner(tunerCallback)
  private val tuner = TunerNew(tunerCallback)

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    context = binding.applicationContext
    methodChannel = MethodChannel(binding.binaryMessenger, "flutter_tuner")
    eventChannel = EventChannel(binding.binaryMessenger, "flutter_tuner_stream")

    methodChannel.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
      try {
          when (call.method) {
            "startTuning" -> {
              if (ContextCompat.checkSelfPermission(context, Manifest.permission.RECORD_AUDIO) != PackageManager.PERMISSION_GRANTED) {
                result.error("401", "Microphone permission not granted", null)
                return
              }
              startTuning()
              result.success(null)
            }
            "stopTuning" -> {
              stopTuning()
              result.success(null)
            }
            else -> result.notImplemented()
          }
      } catch (e: Exception) {
        Log.e(logTag, "Unexpected exception", e)
        result.error(
          "317",
          "Unexpected exception", e.localizedMessage
        )
      }
  }

  private fun startTuning() {
    tuner.start()
  }

  private fun stopTuning() {
    tuner.stop()
  }


  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    eventSink = events
  }

  override fun onCancel(arguments: Any?) {
    eventSink = null
    stopTuning()
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    methodChannel.setMethodCallHandler(null)
    eventChannel.setStreamHandler(null)
  }
}